src/common/threading_utils.cc

```diff
@@ -0,0 +1,47 @@
+/*!
+ * Copyright 2022 by XGBoost Contributors
+ */
+#include "threading_utils.h"
+#if defined(__linux__)
+#include <fcntl.h>
+#include <unistd.h>
+#endif  // defined(__linux__)
+#include "xgboost/logging.h"
+
+namespace xgboost {
+namespace common {
+/**
+ * \brief Get thread limit from CFS
+ *
+ * Modified from
+ * github.com/psiha/sweater/blob/master/include/boost/sweater/hardware_concurrency.hpp
+ *
+ * MIT License: Copyright (c) 2016 Domagoj Šarić
+ */
+int32_t GetCfsCPUCount() noexcept {
+#if defined(__linux__)
+  // https://bugs.openjdk.java.net/browse/JDK-8146115
+  // http://hg.openjdk.java.net/jdk/hs/rev/7f22774a5f42
+  // RAM limit /sys/fs/cgroup/memory.limit_in_bytes
+  // swap limt /sys/fs/cgroup/memory.memsw.limit_in_bytes
+
+  auto read_int = [](char const* const file_path) noexcept {
+    auto const fd(::open(file_path, O_RDONLY, 0));
+    if (fd == -1) {
+      return -1;
+    }
+    char value[64];
+    CHECK(::read(fd, value, sizeof(value)) < signed(sizeof(value)));
+    return std::atoi(value);
```

Comment:
Can we use `std::stoi` here instead? It will throw an exception upon encountering an invalid value. Is this line part of a performance-critical path?

`std::atoi` may have an [undefined behavior](https://en.cppreference.com/w/cpp/string/byte/atoi):
> If no conversion can be performed, ​0​ is returned.
> If the value of the result cannot be represented, i.e. the converted value falls out of range of the corresponding return type, behavior is undefined.

Reply:
```suggestion
    try {
        return std::stoi(value);
    } catch (std::exception&) {
        return -1;
    }
```

Reply:
Thank you for sharing the reference.  Changed to stoi.